Proposed product descriptions/Metrics Improvements.md

```diff
@@ -0,0 +1,184 @@
+
+This project seeks to improve the internal Metrics Service in terms of performance, reliability, maintainability, configurability, and available tools.
+
+The Metrics Service consists of the following components:
+
+1. service instrumentation and data recording
+2. data harvesting
+3. data preparation and normalization
+4. user interface
+5. data extraction
+6. data presentation
+7. public data
+
+This project addresses components 2-6.
+
+#### Existing Services
+
+We currently have one set of scripts in control of data harvesting (2), running inside the kbase infrastructure since they require direct access to KBase databases.
+
+The first version of the metrics service currently handles data preparation (3) as a nightly job and the user interface (4) through a customized Wordpress web site.
+
+Data extraction (5) is provided by a new metrics service which is only used on-demand for reports. Data presentation (6) is provided by importing data extractions into a spreadsheet which has embedded charts and tables.
+
+Public data is provided in the form of a set of json files which are generated by the data preparation scripts (3) and then copied to the production kbase-ui server by a separate process.
+
+#### Improvements and New Services
+
+Data harvesting does not need much additional work, other than simplification. For instance, the extraction of Splunk data has become very slow, but a replacement tool solves this problem. However, we cannot remove the Splunk data extraction because the existing user interface relies upon it.
+
+The rest of the services, which are provided by two sets of scripts, one web site, and a spreadsheet, will be combined into one set of scripts and one web site.
+
+The data preparation and normalization procedures have already been rewritten, but need additional work to support incremental database updating. (It currently must be built from scratch each time.)
+
+The user interface and data presentation (for reports) are currently separate, but would be combined into a single data presentation user interface. The existing web components would be utilized, and additional ones built to replace those in the spreadsheet. Some components, such as the table of user accounts, would be much refactored and improved since the new backing data store is a relational database and not a set of files, and as such naturally provides fast filtering and ordering.
+
+Data extraction already exists in the new metrics service in the form of a REST interface.
+
+The public metrics data would continue to be provided in a similar manner, without an changes required for the ui.
+
+#### Why Now?
+
+Ongoing issues and inefficiencies are a drain on time and resources.
+
+At present, the metrics data gathering process takes over 9 hours to run, which is is potentially affecting production resources (esp. database).
+
```

Comment:
Yes, I didn't tie that in specifically, but the new system does support this much better. Since the data is stored in postgresql, it is much easier to put together a new or modified query along with a new REST end point. It is much better than adding code to grouse through some undocumented JSON structure and then spit out a file. One extension to make this even easier would be to encapsulate the query, REST endpoint, and whatever else is needed into a single file which uses a known interface ... but I think that is probably beyond the scope of this project, or perhaps would be a reach. At the very minimum would be documentation describing how to do this, and at that point it might be just as much work to create an api to simplify the process.